### PR TITLE
V4 candidate 4 remove commit reveal

### PR DESCRIPTION
**Pull request by edmundedgar**
Create the reality.eth v4 version and remove commit-reveal. 

Commit-reveal was intended to protect users against people stealing their answers and front-running them. In practice it was never used in a way where that mattered, ie people have never used it to get questions answered while competing for a bounty, and instead use it to prove things they already think they know. In any case we now have private RPCs and other front-running defences. Removing it makes the code simpler and also makes integrations easier.

This PR removes the commit-reveal functions, then deletes code which can never be run in the absence of a commit-reveal answer. For backwards compatibility it leaves the `is_commitment` parameter in log events and in the hash used to record the record of questions, but always sets it to false.

It should be reasonably simple to check that the ability to supply commit-reveal answers is completely removed, that the other code that is removed would never have run, and that it doesn't make any other changes.

### CODE REVIEW

File: packages/contracts/development/contracts/RealityETH-4.0.sol [hunk 1]
```diff
@@ -440,27 +373,13 @@ contract RealityETH_v3_0 is BalanceHolder, IRealityETH {
...
         address payee;
-        // If the last answer is an unrevealed commit, it's always wrong.
-        // For anything else, the last answer was set as the "best answer" in submitAnswer or submitAnswerReveal.
-        if (is_commitment && !commitments[last_answer_or_commitment_id].is_revealed) {
-            if (commitments[last_answer_or_commitment_id].reveal_ts >= uint32(block.timestamp)) revert YouMustWaitForTheRevealDeadlineBeforeFinalizing();
-            payee = payee_if_wrong;
-        } else {
-            payee = (questions[question_id].best_answer == answer) ? last_answerer : payee_if_wrong;
-        }
+        payee = (questions[question_id].best_answer == answer) ? last_answerer : payee_if_wrong;
```
Comment (josojo) [inline packages/contracts/development/contracts/RealityETH-4.0.sol:382]: could be merged with upper line

File: packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol [hunk 2]
```diff
@@ -526,27 +454,13 @@ contract RealityETH_ERC20_v3_0 is BalanceHolder_ERC20, IRealityETH_ERC20 {
...
         address payee;
-        // If the last answer is an unrevealed commit, it's always wrong.
-        // For anything else, the last answer was set as the "best answer" in submitAnswer or submitAnswerReveal.
-        if (is_commitment && !commitments[last_answer_or_commitment_id].is_revealed) {
-            if (commitments[last_answer_or_commitment_id].reveal_ts >= uint32(block.timestamp)) revert YouMustWaitForTheRevealDeadlineBeforeFinalizing();
-            payee = payee_if_wrong;
-        } else {
-            payee = (questions[question_id].best_answer == answer) ? last_answerer : payee_if_wrong;
-        }
+        payee = (questions[question_id].best_answer == answer) ? last_answerer : payee_if_wrong;
```
Comment (josojo) [inline packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol:463]: same as above

File: packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol [hunk 3]
```diff
@@ -693,16 +607,16 @@ contract RealityETH_ERC20_v3_0 is BalanceHolder_ERC20, IRealityETH_ERC20 {
...
+    /// @param bonds Last-to-first, the bond supplied with each answer
+    /// @param answers Last-to-first, each answer supplied
     function claimWinnings(bytes32 question_id, bytes32[] memory history_hashes, address[] memory addrs, uint256[] memory bonds, bytes32[] memory answers) public stateFinalized(question_id) {
         if (history_hashes.length == 0) revert AtLeastOneHistoryHashEntryMustBeProvided();
 
         // These are only set if we split our claim over multiple transactions.
         address payee = question_claims[question_id].payee;
         uint256 last_bond = question_claims[question_id].last_bond;
-        uint256 queued_funds = question_claims[question_id].queued_funds;
+        uint256 queued_funds = 0;
```
Comment (josojo) [inline packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol:619]: why do you introduce this change? Donät you allow splitting it over several txs

File: packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol [hunk 4]
```diff
@@ -734,16 +647,12 @@ contract RealityETH_ERC20_v3_0 is BalanceHolder_ERC20, IRealityETH_ERC20 {
...
 
-            // If we know who to pay we can go ahead and pay them out, only keeping back last_bond
-            // (We always know who to pay unless all we saw were unrevealed commits)
-            if (payee != NULL_ADDRESS) {
-                _payPayee(question_id, payee, queued_funds);
-                queued_funds = 0;
-            }
+            // Pay out the latest payee, only keeping back last_bond which the next may have a claim on
+            _payPayee(question_id, payee, queued_funds);
+            queued_funds = 0;
```
Comment (josojo) [inline packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol:652]: I think we don't need to reset the variable. It is no longer relevant, right?

File: packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol [hunk 5]
```diff
@@ -712,11 +626,10 @@ contract RealityETH_ERC20_v3_0 is BalanceHolder_ERC20, IRealityETH_ERC20 {
 
         uint256 i;
         for (i = 0; i < history_hashes.length; i++) {
-            // Check input against the history hash, and see which of 2 possible values of is_commitment fits.
-            bool is_commitment = _verifyHistoryInputOrRevert(last_history_hash, history_hashes[i], answers[i], bonds[i], addrs[i]);
+            _verifyHistoryInputOrRevert(last_history_hash, history_hashes[i], answers[i], bonds[i], addrs[i]);
 
             queued_funds = queued_funds + last_bond;
```
Comment (josojo) [inline packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol:631]: I think we can remove queued_funds here

**Review by josojo — APPROVED** [review]
Very nice.

just minor comments.

Since its just taking out code, this might not even an audit.

For long-term maintainablity, I would not keep the unused code changes and just removing all historic code depth. But this is up to you.

A small risk is that you are reducing your testing coverage with this PR. But this is hard for me to evalulate, since I have not dived deep enough into the tests. If you really consider it a risk, we chat about it.

File: packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol (hunk 4 above)
Comment (edmundedgar) [reply packages/contracts/development/contracts/RealityETH_ERC20-4.0.sol]: Nice catch, removed (with no bytecode changes!)

File: packages/contracts/tests/python/test.py [hunk 6]
```diff
@@ -646,6 +661,10 @@ def test_arbitrator_answering_assigning_answerer_wrong_commit(self):
     @unittest.skipIf(WORKING_ONLY, "Not under construction")
     def test_arbitrator_answering_assigning_answerer_wrong(self):
 
+        if not IS_COMMIT_REVEAL_SUPPORTED:
+            print("Skipping test_bond_claim_after_reveal_fail, contract does not support commit-reveal")
```
[the remark this answers is not in the document]
Comment (edmundedgar) [reply packages/contracts/tests/python/test.py]: Nice catch, that wasn't a commit-reveal test and shouldn't have been deactivated.

**Pull request merged**